transaction/std/src/memo_builder/gift_code_sender_memo_builder.rs

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Defines the Memo Builder for the gift code sender memo (0x0002)
+//! specified in MCIP #32
+
+use crate::{
+    memo::{GiftCodeSenderMemo, UnusedMemo},
+    MemoBuilder, ReservedDestination,
+};
+use mc_account_keys::PublicAddress;
+use mc_transaction_core::{Amount, MemoContext, MemoPayload, NewMemoError};
+
+/// There are three possible gift code memo types specified in MCIP #32
+/// | Memo type bytes | Name                        |
+/// | -----------     | -----------                 |
+/// | -->0x0002<--    | Gift Code Sender Memo       |
+/// |    0x0201       | Gift Code Funding Memo      |
+/// |    0x0202       | Gift Code Cancellation Memo |
+/// This memo builder builds a gift code sender memo (0x0002). A gift code
+/// considered redeemed when the Receiver uses the TxOut spend private key
+/// of the gift code TxOut they received from the Sender to send the TxOut
+/// from the sender's gift code subaddress to their own change subaddress.
+/// The destination memo is written into that TxOut at the change address
+/// and includes an optional Utf-8 note up to 64 bytes long the Receiver
+/// can use to record any information they desire about the gift code.
+#[derive(Clone, Debug)]
+pub struct GiftCodeSenderMemoBuilder {
+    // Utf-8 formatted note within the memo that can be up to 64 bytes long
+    note: String,
+    // Whether or not to enable change memos
+    gift_code_change_memo_enabled: bool,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+    // Whether our last note setting attempt was invalid and the note hasn't
+    // been reset with a valid note or cleared
+    attempted_invalid_note: bool,
+}
+
+// Create an empty GiftCodeSenderMemoBuilder
+impl Default for GiftCodeSenderMemoBuilder {
+    fn default() -> Self {
+        Self {
+            note: "".into(),
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+            attempted_invalid_note: false,
+        }
+    }
+}
+
+impl GiftCodeSenderMemoBuilder {
+    /// Set a utf-8 note (up to 64 bytes) onto the sender memo indicating
+    /// any desired info about the gift code. This method will enforce the
+    /// 64 byte limit with a NewMemoErr if the &str passed is longer than
+    /// 64 bytes.
+    pub fn set_gift_code_sender_note(&mut self, note: &str) -> Result<(), NewMemoError> {
+        if note.len() > GiftCodeSenderMemo::MEMO_DATA_LEN {
+            self.attempted_invalid_note = true;
+            return Err(NewMemoError::BadInputs(
+                "Sender note cannot be longer than 64 bytes".into(),
+            ));
+        }
+        self.note = note.into();
+        self.attempted_invalid_note = false;
+        Ok(())
+    }
+    /// Clear the gift code sender note
+    pub fn clear_sender_note(&mut self) {
```

Comment:
IMO, the API is simpler if it's like this:

* `Default` is not implemented
* A `::new` which takes a `String` note is available
* The note provided must be valid, or `::new` returns an error
* We don't need to have `set` and `clear` functions, or track the `attempted_invalid_note` state, because the constructor enforces the invariants.

Reply:
I think that's definitely better, I wrote the transaction builder tests today and I think I wasn't quite seeing how the flow progressed. I definitely agree with this now.

Reply:
In fact I think none of the memo builders actually NEED any get or set methods.